Add --tag-scores option to show tag scores

vbkaisetsu opened the pull request:
This branch adds `--tag-scores` option to show tag scores calculated during the tag prediction.

blocked by #106

kampersanda commented on vaporetto/src/sentence.rs:
```diff
@@ -1089,6 +1095,13 @@ impl<'a, 'b> Sentence<'a, 'b> {
         &mut self.tags
     }
 
+    /// Returns a slice of tag scores.
```
```suggestion
    /// Returns a slice of tag scores.
    ///
    /// The returned slice has the same length as the sentence, and
    /// the `i`-th element corresponds to the `i`-th boundary.
    /// If and only if the `i`-th is a word boundary, it is a `Some` element.
    ///
    /// Each element is a pair as follows:
    ///  * 1st - Tag-string candidates for each tag class.
    ///  * 2nd - Prediction scores for each candidate.
    ///
    /// Let us denote the 1st one by `X[i][j]` and the 2nd one by `Y[k]`.
    /// `Y[k]` stores the score of the `k`-th candidate encountered when
    /// scanning `X` in row-major order, skipping tag classes such that
    /// `X[i]` is of length one.
```
The returned value is very complicated and hard to understand. I think it would be better to define a struct that explains the returned value; however, if you consider this function to be minor and such a struct definition is less important, I will respect that.

In any case, to avoid the cost of understanding this function, its specifications would have to be clarified. Could you clarify the specifications? (My suggestion is one example based on my understanding.)

kampersanda commented on vaporetto/src/sentence.rs:
```diff
@@ -1089,6 +1095,13 @@ impl<'a, 'b> Sentence<'a, 'b> {
 
+    /// Returns a slice of tag scores.
+    #[allow(clippy::type_complexity)]
+    #[inline]
```
```suggestion
    #[doc(hidden)]
    #[inline]
```
If we do not introduce an explanatory struct for the returned value, I think it would be better to hide it from the documentation. It would be difficult for anyone other than contributors to use this function.

kampersanda commented on predict/src/main.rs:
```diff
@@ -70,6 +74,56 @@ fn print_scores(s: &Sentence, mut out: impl Write) -> Result<(), Box<dyn std::er
+    {
+        write!(out, "{c}")?;
+        if *b == CharacterBoundary::WordBoundary {
+            if let Some((cands, scores)) = s.as_ref() {
```
Is `None` possible?

vbkaisetsu replied:
It becomes None if the token is unknown

vbkaisetsu commented:
I removed `Sentence::tag_scores()` and added `Token::tag_candidates()`.
The previous functions relied strongly on implementation details, but the new function formats the array and provides intuitive output.

kampersanda commented on vaporetto/src/sentence.rs:
```diff
@@ -1203,6 +1209,31 @@ impl<'a, 'b> Token<'a, 'b> {
+    /// is a tuple of the tag name and its score.
+    pub fn tag_candidates(&self) -> Vec<Vec<(&'b str, i32)>> {
+        let mut results = vec![];
+        if let Some((tags, scores)) = self.sentence.tag_scores[self.end - 1].as_ref() {
```
If `Predictor::store_tag_scores` is unset, this line will panic implicitly due to "index out of bounds". Could you address it?

kampersanda replied:
@vbkaisetsu 

How about this suggestion?

More precisely, I mean that:

- `Token::tag_candidates` assumes that `self.sentence.tag_scores` is not empty.
- However, this instance is resized only when `Predictor::store_tag_scores` is set ([here](https://github.com/daac-tools/vaporetto/blob/tag-scores/vaporetto/src/predictor.rs#L562-L564)).
- If `Predictor::store_tag_scores` is unset, `Token::tag_candidates` will always panic implicitly because `self.sentence.tag_scores` is empty.

I think we should pre-check if `tag_scores` is empty or not and return None if empty.

vbkaisetsu replied:
I added the Panics section in the previous commit.

I don't want to use Result here.
If the function returns `Result::Err`, it basically means that the conditions for normal completion have not been met at some point that cannot be controlled by the program, such as user input or external state.
This function should always return a normal value when `store_tag_scores()` is `true`, and I think it should panic instead of returning `Result` since this function should not be called when the flag is `false`.

kampersanda replied:
I overlooked the Panics section in the previous commit. Sorry.

I agree with your opinion.

kampersanda commented on vaporetto/src/sentence.rs:
```diff
@@ -1203,6 +1209,31 @@ impl<'a, 'b> Token<'a, 'b> {
+    /// The return value is a two-dimensional array. The outer array index corresponding to the
+    /// return value of [`Token::tags()`]. The inner array is a candidate set, where each element
+    /// is a tuple of the tag name and its score.
+    pub fn tag_candidates(&self) -> Vec<Vec<(&'b str, i32)>> {
```
```suggestion
    #[cfg(feature = "tag-prediction")]
    pub fn tag_candidates(&self) -> Vec<Vec<(&'b str, i32)>> {
```
I guess that this function is never used unless`tag-prediction` is specified. The same is true for `Token::tags`.

kampersanda replied:
I see

kampersanda reviewed (commented):
(dummy)
